src/functional/swing_slide_filter.zig

```diff
@@ -0,0 +1,411 @@
+// Copyright 2024 TerseTS Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of The Swing Filter and Slide Filter algorithms from the paper
+//! "Hazem Elmeleegy, Ahmed K. Elmagarmid, Emmanuel Cecchet, Walid G. Aref, and Willy Zwaenepoel.
+//! Online piece-wise linear approximation of numerical streams with precision guarantees.
+//! Proc. VLDB Endow. 2, 1, 2009.
+//! https://doi.org/10.14778/1687627.1687645".
+
+const std = @import("std");
+const math = std.math;
+const mem = std.mem;
+const testing = std.testing;
+const ArrayList = std.ArrayList;
+
+const tersets = @import("../tersets.zig");
+const Error = tersets.Error;
+
+/// Segment represented by the points: (start_time, start_value) and (end_time, end_value).
+/// The `start_time` and `end_time` represent indices of the time series.
+const Segment = struct {
+    start_time: usize,
+    start_value: f64,
+    end_time: usize,
+    end_value: f64,
+};
+
+/// LinearFunction represented by the function y = slope*x+intercept.
+/// It uses f80 to improve numerical stability.
+const LinearFunction = struct {
+    slope: f80,
+    intercept: f80,
+};
+
+/// Converts `value`:usize to f80
+fn usizeToF80(value: usize) f80 {
+    return @as(f80, @floatFromInt(value));
+}
+
+/// Computes the intercept coefficient of a line that passes through the point
+/// (x_time, y_value) and with the given slope coefficient.
+fn computeInterceptCoefficient(slope: f80, x_time: usize, y_value: f64) f80 {
+    return y_value - slope * usizeToF80(x_time);
+}
+
+/// Updates the linear function coeficients in `line` that passes throught the two points
+/// of the `segment`. The line is swinged down or up based on the `error_bound`.
+/// If `error_bound` is negative, `line` is swing down. It is swing up otherwise.
+fn updateSwingLine(segment: Segment, line: *LinearFunction, error_bound: f32) void {
+    if (segment.end_time != segment.start_time) {
+        const duration: f80 = @floatFromInt(segment.end_time - segment.start_time);
+        line.slope = (segment.end_value + error_bound - segment.start_value) / duration;
+        line.intercept = segment.start_value - line.slope * usizeToF80(segment.start_time);
+    } else {
+        line.slope = 0.0;
+        line.intercept = segment.start_value;
+    }
+}
+
+/// Evaluate the linear function `line` at `x_time`.
```

Comment:
Unclear if it should be "at `x_time`" or "with `x_time`"

Reply:
Fixed. https://github.com/cmcuza/TerseTS/pull/13#discussion_r1618911949